presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/trait/CardinalityTraitCalculationRuleSet.java

```diff
@@ -0,0 +1,152 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public final class CardinalityTraitCalculationRuleSet
+{
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                getEnforceSingleRowCardinalityCalculationRule(),
+                getValuesCardinalityCalculationRule(),
+                getAggregationCardinalityCalculationRule(),
+                getProjectCardinalityCalculationRule(),
+                getMarkDistinctCardinalityCalculationRule(),
+                getExchangeCardinalityCalculationRule(),
+                getFilterCardinalityCalculationRule(),
+                getLimitCardinalityCalculationRule());
+    }
+
+    public Rule<ProjectNode> getProjectCardinalityCalculationRule()
+    {
+        return propagateTrait(project());
+    }
+
+    public Rule<EnforceSingleRowNode> getEnforceSingleRowCardinalityCalculationRule()
+    {
+        return applyTraitRule(enforceSingleRow(), node -> scalar());
+    }
+
+    public Rule<ValuesNode> getValuesCardinalityCalculationRule()
+    {
+        return applyTraitRule(values(), node -> exactly(node.getRows().size()));
+    }
+
+    public Rule<AggregationNode> getAggregationCardinalityCalculationRule()
+    {
+        return calculateCardinalityRule(aggregation(), (node, context) -> {
+            if (node.getStep() == SINGLE && node.getGroupingSets().size() == 1 && node.hasEmptyGroupingSet()) {
+                return Result.setTrait(scalar());
+            }
+            return Result.empty();
+        });
+    }
+
+    public Rule<MarkDistinctNode> getMarkDistinctCardinalityCalculationRule()
+    {
+        return propagateTrait(markDistinct());
+    }
+
+    public Rule<ExchangeNode> getExchangeCardinalityCalculationRule()
+    {
+        return calculateCardinalityRule(exchange(), (node, context) -> {
+            if (node.getSources().size() != 1) {
+                return Result.empty();
+            }
+            Optional<CardinalityTrait> sourceCardinality = getCardinality(context, getOnlyElement(node.getSources()));
+            return sourceCardinality.map(Result::setTrait)
+                    .orElseGet(Result::empty);
+        });
+    }
+
+    public Rule<FilterNode> getFilterCardinalityCalculationRule()
+    {
+        return calculateCardinalityRule(filter(), (node, context) -> {
+            Optional<CardinalityTrait> sourceCardinality = getCardinality(context, getOnlyElement(node.getSources()));
+            return sourceCardinality.map(cardinalityTrait -> Result.setTrait(atMost(cardinalityTrait.getMaxCardinality())))
+                    .orElseGet(Result::empty);
+        });
+    }
+
+    public Rule<LimitNode> getLimitCardinalityCalculationRule()
+    {
+        // TODO consider source min cardinality, it requires possibility to updated trait with more specific
+        return applyTraitRule(limit(), node -> atMost(node.getCount()));
```

Comment:
Shouldn't this rule take into account source cardinality? If limit is > maxCardinality, then the result cardinality should match the source, right?

```
    public Rule<LimitNode> getLimitCardinalityCalculationRule()
    {
        return calculateCardinalityRule(limit(), (node, context) -> {
            Optional<CardinalityTrait> sourceCardinality = getCardinality(context, getOnlyElement(node.getSources()));
            long limitCount = node.getCount();
            return sourceCardinality.map(sourceTrait -> Result.ofTrait(new CardinalityTrait(Math.min(sourceTrait.getMinCardinality(), limitCount), Math.min(sourceTrait.getMaxCardinality(), limitCount))))
                    .orElse(Result.ofTrait(atMost(limitCount)));
        });
    }
```

Reply:
Yes. Please see the the `TODO` above.